lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -832,5 +889,25 @@ public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
       return new DefaultBulkScorer(new ConstantScoreScorer(this, 0f, ScoreMode.COMPLETE_NO_SCORES, disi));
     }
 
+    // Perform a cache load asynchronously
+    private void cacheAsynchronously(LeafReaderContext context, IndexReader.CacheHelper cacheHelper) {
+      /*
+       * If the current query is already being asynchronously cached,
+       * do not trigger another cache operation
+       */
+      if (inFlightAsyncLoadQueries.add(in.getQuery()) == false) {
+        return;
+      }
+
+      FutureTask<Void> task = new FutureTask<>(() -> {
+        DocIdSet localDocIdSet = cache(context);
+        putIfAbsent(in.getQuery(), localDocIdSet, cacheHelper);
+
+        //remove the key from inflight -- the key is loaded now
+        inFlightAsyncLoadQueries.remove(in.getQuery());
+        return null;
+      });
+      executor.execute(task);
```

Comment:
That depends on the implementation of the executor passed in -- I think we do not publish any such contract in `IndexSearcher`, and since this is reusing the same `Executor`, the behaviour is the same.

Reply:
Well, the `RejectedExecutionException` is a public part of this API ... I think we need to handle it in some fashion, maybe fallback to sync'd (calling-thread-executes) implementation, or, we skip caching and continue running the query uncached?

Reply:
Fixed. I think that actually is a valid problem even in `IndexSearcher`? I will open another issue to track that.

Reply:
Yeah it is, thanks for opening separate issue ...